Ensure that we fakeify tensor subclasses when they are initially tracked

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90295
* #90286
* #90202
* #90186
* #90185
* #90179
* __->__ #90009
* #89962

The old code didn't actually fakeify traceable tensor subclasses at the
time they are added as a GraphArg to the module; now we do, by ignoring
the subclass during fakeification and relying on Dynamo to simulate
the subclass on top.  See comments for more details.

BTW, this codepath is super broken, see filed issues linked on the
inside.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

ezyang committed:
Ensure that we fakeify tensor subclasses when they are initially tracked

The old code didn't actually fakeify traceable tensor subclasses at the
time they are added as a GraphArg to the module; now we do, by ignoring
the subclass during fakeification and relying on Dynamo to simulate
the subclass on top.  See comments for more details.

BTW, this codepath is super broken, see filed issues linked on the
inside.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Ensure that we fakeify tensor subclasses when they are initially tracked"


The old code didn't actually fakeify traceable tensor subclasses at the
time they are added as a GraphArg to the module; now we do, by ignoring
the subclass during fakeification and relying on Dynamo to simulate
the subclass on top.  See comments for more details.

BTW, this codepath is super broken, see filed issues linked on the
inside.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

albanD commented on torch/_dynamo/variables/builder.py:
```diff
@@ -550,40 +550,65 @@ def wrap_tensor(self, value: torch.Tensor):
+            # Ordinarily, we would fakeify a tensor so that it can get dynamic
+            # shapes and be computed on without triggering actual operations.
+            # However, how can we fakeify a tensor subclass?  Ordinary
+            # inheritance (nor multiple inheritance) won't work work.
```
nit double work

albanD commented on torch/_dynamo/variables/builder.py:
```diff
@@ -550,40 +550,65 @@ def wrap_tensor(self, value: torch.Tensor):
-                    subclass_type,
-                )
-            return tensor_variable
+            assert type(value) in (torch.Tensor, torch.nn.Parameter)
```
nit: this will fail for every unknown subclasses so we should have a nicer error message here as this will be quite user facing no?

ezyang replied:
No, we test if it's a traceable tensor type earlier in dynamo, so if you have gotten here we've messed up the preconditions.

albanD reviewed (commented):
Sounds ok but we should test that what we want to work does work?

wconstab reviewed (commented):
idk, seems directionally good, does it help unblock some current bugs or is it really WIP and needs more before its useful?

ezyang commented:
@wconstab yeah, this is part of a larger workstream of guaranteeing that we never allocate SymInts for intermediates. To do this, I need to not be fakeifying intermediates (since fakeifying is what causes SymInts to get allocated). This is one of the unit tests that triggered when tightening the invariants here.

ezyang commented:
> Sounds ok but we should test that what we want to work does work?

There are tensor subclass tests which were added when this feature was initially added, and they are still passing after this change. If you like, I could add tests for more realistic tensor subclasses and xfail them (since they aren't going to work, the implementation is just not good enough), but I figured it would probably be better for whoever actually is going to implement this to do it.

wconstab reviewed (approved):
ok, stamped.

voznesenskym commented:
Does this make `test/dynamo/test_modules.py`'s `test_simple_torch_function` pass?

ezyang commented:
The xfail did not appear to start passing

ezyang committed:
Update on "Ensure that we fakeify tensor subclasses when they are initially tracked"


The old code didn't actually fakeify traceable tensor subclasses at the
time they are added as a GraphArg to the module; now we do, by ignoring
the subclass during fakeification and relying on Dynamo to simulate
the subclass on top.  See comments for more details.

BTW, this codepath is super broken, see filed issues linked on the
inside.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang committed:
Update on "Ensure that we fakeify tensor subclasses when they are initially tracked"


The old code didn't actually fakeify traceable tensor subclasses at the
time they are added as a GraphArg to the module; now we do, by ignoring
the subclass during fakeification and relying on Dynamo to simulate
the subclass on top.  See comments for more details.

BTW, this codepath is super broken, see filed issues linked on the
inside.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented:
Turns out when I rebased this PR, the test is now passing! Hooray!

ezyang committed:
Update on "Ensure that we fakeify tensor subclasses when they are initially tracked"


The old code didn't actually fakeify traceable tensor subclasses at the
time they are added as a GraphArg to the module; now we do, by ignoring
the subclass during fakeification and relying on Dynamo to simulate
the subclass on top.  See comments for more details.

BTW, this codepath is super broken, see filed issues linked on the
inside.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge -f "all green merge bot is broken"
